### PR TITLE
[doc] auto-generation of class apis

### PR DESCRIPTION
**Pull request by can-anyscale**
After discuss offline with @bveeramani and @angelinalg, this is another attempt to auto-generate doc apis. Compared to the previous work, this focuses only on a single class and its methods. It also supports the common pattern of logical group of methods within a class, which can be documented in the code itself.

Only the content of the class doc are generated, and users have freedom to include that content in any structure they want. 

The new flow for adding a new class will be:
- Adding the class name into `_autogen.rst`
- Include the class full name, for example, `ray.data.DataIterator` in any top level rst.

I made an demo in this PR using DataIterator class.

Test:
- CI
- https://anyscale-ray--46243.com.readthedocs.build/en/46243/

### CODE REVIEW

File: doc/source/_templates/autosummary/class_v2.rst [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+.. autoclass:: {{ objname }}
+   :noindex:
+
+{% block methods %}
+{% if methods %}
+{% set api_groups = methods | get_api_groups(name, module) %}
+{% for api_group in api_groups %}
+
+{% if api_groups | length > 1 %}
```
Comment (can-anyscale) [inline doc/source/_templates/autosummary/class_v2.rst:13]: this skips the logical group if there is only one

**Review by bveeramani — APPROVED** [review]
Nice

File: python/ray/data/iterator.py [hunk 2]
```diff
@@ -102,6 +102,7 @@ def _to_block_iterator(
         """
         raise NotImplementedError
 
+    @PublicAPI(stability="beta")
```
Comment (aslonnie) [inline python/ray/data/iterator.py:105]: maybe separate these annotations in another PR? it is not clear if these changes are related to the auto gen attempt

File: doc/source/conf.py [hunk 3]
```diff
@@ -1,15 +1,19 @@
...
 from jinja2.filters import FILTERS
 import sphinx
 from sphinx.ext import autodoc
+from sphinx.ext.autosummary import generate
+from sphinx.util.inspect import safe_getattr
 from docutils import nodes
 import pathlib
 import logging
 
+DEFAULT_API_GROUP = "others"
```
Comment (aslonnie) [inline doc/source/conf.py:15]: nit: `_DEFAULT_API_GROUP` ?

maybe use value `"_"` or "default" or "na" or "none"? "others" feel a bit weird as a default string.. or maybe it is commonly used in the docs world?

image

File: doc/source/conf.py [hunk 4]
```diff
@@ -385,7 +397,40 @@ def filter_out_undoc_class_members(member_name, class_name, module_name):
         return ""
 
 
+def get_api_groups(method_names, class_name, module_name):
```
Comment (aslonnie) [inline doc/source/conf.py:400]: nit: `_get_api_group`? seems that they are just going to be assigned to the `FILTERS`?

**Review by aslonnie — APPROVED** [review]

**Comment by aslonnie** [comment]
hmm .. just got auto merged .. anyways..

**Comment by can-anyscale** [comment]
i'll follow up in another pr ;)

File: doc/source/conf.py (hunk 3 above)
Comment (can-anyscale) [reply doc/source/conf.py:15]: this word is rendered as a section name in the template (class_v2.rst) so i choose something readable
